pennylane/devices/default_mixed.py

```diff
@@ -116,8 +153,129 @@ def _get_kraus(self, operation):  # pylint: disable=no-self-use
+        new_col_indices = ABC[
+            2 * self.num_wires + len(channel_wires) : 2 * (self.num_wires + len(channel_wires))
+        ]
+
+        # index for summation over Kraus operators
+        kraus_index = ABC[
+            2 * (self.num_wires + len(channel_wires)) : 2 * (self.num_wires + len(channel_wires))
+            + 1
+        ]
+
+        # new state indices replace row and column indices with new ones
+        new_state_indices = functools.reduce(
+            lambda old_string, idx_pair: old_string.replace(idx_pair[0], idx_pair[1]),
+            zip(col_indices + row_indices, new_col_indices + new_row_indices),
+            state_indices,
+        )
+
+        # index mapping for einsum, e.g., 'iga,abcdef,idh->gbchef'
+        einsum_indices = (
+            "{kraus_index}{new_row_indices}{row_indices}, {state_indices},"
+            "{kraus_index}{col_indices}{new_col_indices}->{new_state_indices}".format(
+                kraus_index=kraus_index,
+                new_col_indices=new_col_indices,
+                col_indices=col_indices,
+                state_indices=state_indices,
+                row_indices=row_indices,
+                new_row_indices=new_row_indices,
+                new_state_indices=new_state_indices,
+            )
+        )
+
+        self._state = self._einsum(einsum_indices, kraus, self._state, kraus_dagger)
+
+    def _apply_diagonal_unitary(self, eigvals, wires):
+        r"""Apply a diagonal unitary gate specified by a list of eigenvalues. This method uses
+        the fact that the unitary is diagonal for a more efficient implementation.
+
+        Args:
+            eigvals (array): eigenvalues (phases) of the diagonal unitary
+            wires (Wires): target wires
+        """
+
+        channel_wires = self.map_wires(wires)
+
+        # reshape vectors
+        eigvals = self._cast(self._reshape(eigvals, [2] * len(channel_wires)), dtype=self.C_DTYPE)
+
+        # Tensor indices of the state. For each qubit, need an index for rows *and* columns
+        state_indices = ABC[: 2 * self.num_wires]
+
+        # row indices of the quantum state affected by this operation
+        row_wires_list = channel_wires.tolist()
+        row_indices = "".join(ABC_ARRAY[row_wires_list].tolist())
+
+        # column indices are shifted by the number of wires
+        col_wires_list = [w + self.num_wires for w in row_wires_list]
+        col_indices = "".join(ABC_ARRAY[col_wires_list].tolist())
+
+        einsum_indices = "{row_indices},{state_indices},{col_indices}->{state_indices}".format(
+            col_indices=col_indices, state_indices=state_indices, row_indices=row_indices
+        )
+
+        self._state = self._einsum(einsum_indices, eigvals, self._state, self._conj(eigvals))
+
+    def _apply_operation(self, operation):
+        """Applies operations to the internal device state.
+
+        Args:
+            operation (~.Operation): operation to apply on the device
```

Comment:
Just noticed this --- when specifying types for Sphinx, there is no need to include the `~`. In fact, this will cause Sphinx to fail to link this to the class docstring page!

There might be other cases where the `~` needs to be removed above and below

```suggestion
            operation (.Operation): operation to apply on the device
```